src/hotspot/share/oops/constantPool.cpp

```diff
@@ -1276,6 +1276,33 @@ void ConstantPool::unreference_symbols() {
   }
 }
 
+// Returns true if the current mismatch is due to a resolved/unresolved
+// class pair. Otherwise, returns false.
+bool ConstantPool::is_unresolved_class_mismatch(int index1,
+  const constantPoolHandle& cp2, int index2) {
+
+  jbyte t1 = tag_at(index1).value();
+  if (t1 != JVM_CONSTANT_Class && t1 != JVM_CONSTANT_UnresolvedClass) {
+    return false;  // wrong entry type; not our special case
+  }
+
+  jbyte t2 = cp2->tag_at(index2).value();
+  if (t2 != JVM_CONSTANT_Class && t2 != JVM_CONSTANT_UnresolvedClass) {
+    return false;  // wrong entry type; not our special case
+  }
+
+  if (t1 == t2) {
+    return false;  // not a mismatch; not our special case
+  }
+
+  char *s1 = klass_name_at(index1)->as_C_string();
+  char *s2 = cp2->klass_name_at(index2)->as_C_string();
+  if (strcmp(s1, s2) != 0) {
+    return false;  // strings don't match; not our special case
+  }
```

Comment:
klass_name_at() returns a Symbol* that's in the SymbolTable so this never needed to do a strcmp.  It can test for equality.